DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/src/svd.hpp

```diff
@@ -0,0 +1,273 @@
+#include "tuple.hpp"
+
+// kernels
+class IDInputMatrixFromDDRToLocalMem;
+class IDInputTransMatrixFromDDRToLocalMem;
+class IDCovarianceMatrixComputation;
+class IDEigenValuesAndVectorsComputation;
+class IDInputMatrixFromDDRToLocalMemPostProcess;
+class IDUSVFromEigens;
+class IDOrthogonalizingU;
+class IDRankDeficientFlagFromLocalMemToDDR;
+class IDUMatrixFromLocalMemToDDR;
+class IDRMatrixFromLocalMemToDDR;
+class IDSMatrixFromLocalMemToDDR;
+class IDVMatrixFromLocalMemToDDR;
+
+// pipes
+class CMP;
+class IMP;
+class IMP2;
+class ITMP;
+class IDP;
+class EValP;
+class EVecP;
+class RDFP;
+class OUTMP;
+class OUMP;
+class ORMP;
+class OSMP;
+class OVMP;
+
+template <unsigned rows,            // row size of input matrix
+          unsigned cols,            // col size of input matrix
+          unsigned k_fixed_iteration, // fixed number of QR iteration
+          unsigned k_raw_latency,   // RAW latency for triangular loop
+                                    // optimization in the QR iteration
+          int k_zero_threshold_1e,  // Threshold from which we consider a
+          typename T                // The datatype for the computation
+          >
+double singularValueDecomposition(
+    std::vector<T> &a_matrix,  // input matrix A (col maj)
+    std::vector<T> &u_matrix,  // left singular vectors, U (col maj)
+    std::vector<T> &s_matrix,  // list of singular values, S
+    std::vector<T> &v_matrix,  // right singular vectors, V (row maj)
+    std::vector<ac_int<1, false>>
+        &rank_deficient_flag,  // Output a flag that is 1 if the input matrix
+                               // is considered rank deficient
+    sycl::queue &q, int repetitions) {
+  constexpr int kNumElementsPerDDRBurst = 8;            // total of 256 bit
+  constexpr int kInputMatrixSize = rows * cols;         // row x col
+  constexpr int kEigenVectorsMatrixSize = cols * cols;  // col x col
+  constexpr int kUMatrixSize = rows * rows;
+  constexpr int kRMatrixSize = rows * (rows + 1) / 2;
+
+  // pipes
+  using PipeType = fpga_tools::NTuple<T, kNumElementsPerDDRBurst>;
+
+  using InputMatrixPipe = sycl::ext::intel::pipe<IMP, PipeType, 16>;
+  using InputMatrixPipe2 = sycl::ext::intel::pipe<IMP2, PipeType, 16>;
+  using CovarianceMatrixPipe = sycl::ext::intel::pipe<CMP, PipeType, 16>;
+  using EigenValuesPipe = sycl::ext::intel::pipe<EValP, T, 3>;
+  using EigenVectorsPipe = sycl::ext::intel::pipe<EVecP, PipeType, 3>;
+  using RankDeficientFlagPipe =
+      sycl::ext::intel::pipe<RDFP, ac_int<1, false>, 3>;
+  using UTempMatrixPipe = sycl::ext::intel::pipe<OUTMP, PipeType, 3>;
+  using UMatrixPipe = sycl::ext::intel::pipe<OUMP, PipeType, 3>;
+  using RMatrixPipe =
+      sycl::ext::intel::pipe<ORMP, T, kNumElementsPerDDRBurst * 4>;
+  using SMatrixPipe = sycl::ext::intel::pipe<OSMP, PipeType, 3>;
+  using VMatrixPipe = sycl::ext::intel::pipe<OVMP, PipeType, 3>;
+
+  // USM allocations
+  T *input_matrix_device;
+  ac_int<1, false> *rank_deficient_flag_device;
```

Comment:
```suggestion
  bool *rank_deficient_flag_device;
```
might as well use a good ol' `bool` here since you will be doing a DRAM transaction anyway.

Reply:
The streamingEigen kernel would need to make the change to 'bool' as well

Reply:
I am having trouble using `q.memcpy` with `bool` type, I will open a case

Reply:
https://hsdes.intel.com/appstore/article/#/22019888605